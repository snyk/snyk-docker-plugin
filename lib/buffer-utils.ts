import * as crypto from "crypto";
import { HashAlgorithm } from "./types";

const HASH_ENCODING = "hex";

export async function bufferToSha1(buffer: Buffer): Promise<string> {
  const hash = crypto.createHash(HashAlgorithm.Sha1);
  const chunkSize = 100 * 1024 * 1024; // 100 MB

  return new Promise((resolve, reject) => {
    try {
      for (let offset = 0; offset < buffer.length; offset += chunkSize) {
        const end = Math.min(offset + chunkSize, buffer.length);
        const chunk = buffer.slice(offset, end);
        hash.update(chunk);
      }

      const digest = hash.digest(HASH_ENCODING);
      resolve(digest);
    } catch (err) {
      reject(err);
    }
  });
}
