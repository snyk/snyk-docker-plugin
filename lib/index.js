var subProcess = require('./sub-process');
var fetchSnykDockerAnalyzer = require('./fetch-snyk-docker-analyzer');
var debug = require('debug')('snyk');

module.exports = {
  inspect: inspect,
};

function inspect(root, targetFile, options) {
  const targetImage = root;
  return fetchSnykDockerAnalyzer()
  .then(function (analyzerBinaryPath) {
    return Promise.all([
      getRuntime(),
      getDependencies(analyzerBinaryPath, targetImage),
    ])
    .then(function (result) {
      var metadata = {
        name: 'snyk-docker-plugin',
        runtime: result[0],
        packageManager: result[1].packageManager,
      };
      var package = result[1].package;
      return {
        plugin: metadata,
        package: package,
      };
    });
  })
}

function getRuntime() {
  return subProcess.execute('docker', ['version'])
  .then(function (output) {
    versionMatch = /Version:\s+(.*)\n/.exec(output);
    if (versionMatch) {
      return 'docker ' + versionMatch[1]
    }
    return undefined;
  });
}

function handleCommonErrors(error, targetImage) {
  if (error.indexOf('command not found') !== -1) {
    throw new Error('Snyk docker CLI was not found')
  }
  if (error.indexOf('Cannot connect to the Docker daemon') !== -1) {
    throw new Error('Cannot connect to the Docker daemon. Is the docker'
    + ' daemon running?')
  }
  if ((error.indexOf('Error loading image from docker engine') !== -1) ||
      (error.indexOf('Error performing image analysis') !== -1)) {
    throw new Error('Docker image was not found: ' + targetImage)
  }
  if (error.indexOf('Error getting docker client:') !== -1) {
    throw new Error('Failed getting docker client')
  }
  if (error.indexOf('Error processing image:') !== -1) {
    throw new Error('Failed processing image:' + targetImage)
  }
}

function getDependencies(analyzerBinaryPath, targetImage) {
  var result;
  return subProcess.execute(
    analyzerBinaryPath,
    buildArgs(targetImage)
  )
  .then(function (output) {
    result = parseAnalysisResults(output);
    return buildTree(
      targetImage, result.type, result.depInfosList, result.osRelease);
  })
  .then(function (package) {
    return {
      package: package,
      packageManager: result.type,
    };
  })
  .catch(function (error) {
    if (typeof error === 'string') {
      debug(`Error while running analyser: '${error}'`);
      handleCommonErrors(error, targetImage);
      errorMsg = error;
      errorMatch = /msg="(.*)"/g.exec(errorMsg)
      if (errorMatch) {
        errorMsg = errorMatch[1]
      }
      throw new Error(errorMsg);
    }

    throw error;
  })
}

function parseAnalysisResults(analysisOut) {
  var analysisJson = JSON.parse(analysisOut);

  var analysisResult = analysisJson.results.filter(function (res) {
    return res.Analysis && res.Analysis.length > 0;
  })[0];

  // TODO: if analysisResult is undefined -
  //  throw a nice error, or report 0 deps

  var depType;
  switch (analysisResult.AnalyzeType) {
    case 'Apt': {
      depType = 'deb';
      break;
    }
    default: {
      depType = analysisResult.AnalyzeType.toLowerCase()
    }
  }

  return {
    type: depType,
    depInfosList: analysisResult.Analysis,
    osRelease: analysisJson.osRelease,
  }
}

function buildTree(targetImage, depType, depInfosList, osRelease) {
  var targetSplit = targetImage.split(':');
  var imageName = targetSplit[0];
  var imageVersion = targetSplit[1] ? targetSplit[1] : 'latest';

  var root = {
    name: imageName,
    version: imageVersion,
    dockerOSRelease: osRelease,
    packageFormatVersion: depType + ':0.0.1',
    dependencies: {},
  };

  var depsMap = depInfosList.reduce(function (acc, depInfo) {
    var name = depInfo.Name;
    acc[name] = depInfo;
    return acc;
  }, {});

  var virtualDepsMap = depInfosList.reduce(function (acc, depInfo) {
    var providesNames = depInfo.Provides || [];
    providesNames.forEach(function (name) {
      acc[name] = depInfo;
    });
    return acc;
  }, {});

  var attachDeps = function (depInfos) {
    depInfos.forEach(function (depInfo) {
      var subtree = buildTreeRecurisve(
        depInfo.Name, new Set(), depsMap, virtualDepsMap);
      if (subtree) {
        root.dependencies[subtree.name] = subtree;
      }
    })
  };

  // attach (as direct deps) pkgs not marked auto-installed:
  var manuallyInstalledDeps = depInfosList.filter(function (depInfo) {
    return !depInfo.AutoInstalled;
  })
  attachDeps(manuallyInstalledDeps);

  // attach (as direct deps) pkgs marked as auto-insatalled,
  //  but not dependant upon:
  var notVisitedDeps = depInfosList.filter(function (depInfo) {
    var depName = depInfo.Name;
    return !(depsMap[depName]._visited);
  })
  attachDeps(notVisitedDeps);

  return root;
}

function buildTreeRecurisve(
    depName, ancestors, depsMap, virtualDepsMap) {
  var depInfo = depsMap[depName] || virtualDepsMap[depName];
  if (!depInfo) {
    return null;
  }

  var fullName = depFullName(depInfo);
  if (ancestors.has(fullName)) {
    return null;
  }

  const tree = {
    name: fullName,
    version: depInfo.Version,
  };

  //TODO: if we do this we can avoid tracking ancestors
  if (depInfo._visited) {
    return tree;
  }
  depInfo._visited = true;

  var newAncestors = (new Set(ancestors)).add(fullName);

  var deps = depInfo.Deps || {};
  Object.keys(deps).forEach(function (depName) {
    var subTree = buildTreeRecurisve(
      depName, newAncestors, depsMap, virtualDepsMap);
    if (subTree) {
      if (!tree.dependencies) {
        tree.dependencies = {};
      }
      tree.dependencies[subTree.name] = subTree;
    }
  });

  return tree;
}

function depFullName(depInfo) {
  var fullName = depInfo.Name;
  if (depInfo.Source) {
    fullName = depInfo.Source + '/' + fullName;
  }
  return fullName;
}

function buildArgs(targetImage) {
  var args = ['analyze', targetImage];
  return args;
}
