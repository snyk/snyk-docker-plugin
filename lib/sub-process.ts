import * as childProcess from "child_process";
import { quoteAll } from "shescape/stateless";

export { execute, CmdOutput };
interface CmdOutput {
  stdout: string;
  stderr: string;
}

function execute(
  command: string,
  args: string[],
  options?,
): Promise<CmdOutput> {
  const spawnOptions: any = { shell: true, env: process.env };
  if (options && options.cwd) {
    spawnOptions.cwd = options.cwd;
  }
  args = quoteAll(args, spawnOptions);

  // Before spawning an external process, we look if we need to restore the system proxy configuration,
  // which overides the cli internal proxy configuration.
  if (process.env.SNYK_SYSTEM_HTTP_PROXY !== undefined) {
    spawnOptions.env.HTTP_PROXY = process.env.SNYK_SYSTEM_HTTP_PROXY;
  }
  if (process.env.SNYK_SYSTEM_HTTPS_PROXY !== undefined) {
    spawnOptions.env.HTTPS_PROXY = process.env.SNYK_SYSTEM_HTTPS_PROXY;
  }
  if (process.env.SNYK_SYSTEM_NO_PROXY !== undefined) {
    spawnOptions.env.NO_PROXY = process.env.SNYK_SYSTEM_NO_PROXY;
  }

  return new Promise((resolve, reject) => {
    let stdout = "";
    let stderr = "";

    const proc = childProcess.spawn(command, args, spawnOptions);
    proc.stdout.on("data", (data) => {
      stdout = stdout + data;
    });
    proc.stderr.on("data", (data) => {
      stderr = stderr + data;
    });

    proc.on("close", (code) => {
      const output = { stdout, stderr };
      if (code !== 0) {
        return reject(output);
      }
      resolve(output);
    });
  }) as Promise<CmdOutput>;
}
