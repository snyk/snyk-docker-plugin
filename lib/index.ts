const debug = require('debug')('snyk');

import * as analyzer from './analyzer';
import * as subProcess from './sub-process';
import * as dockerFile from './docker-file';

export {
  inspect,
};

function inspect(root: string, targetFile?: string, options?: any) {
  const targetImage = root;
  return Promise.all([
    getRuntime(),
    getDependencies(targetImage),
    dockerFile.getBaseImageName(targetFile),
  ])
    .then((result) => {
      const metadata = {
        name: 'snyk-docker-plugin',
        runtime: result[0],
        packageManager: result[1].packageManager,
        dockerImageId: result[1].imageId,
      };
      const pkg: any = result[1].package;
      pkg.docker = pkg.docker || {};
      pkg.docker.baseImage = result[2];
      return {
        plugin: metadata,
        package: pkg,
      };
    });
}

function getRuntime() {
  return subProcess.execute('docker', ['version'])
    .then((output) => {
      const versionMatch = /Version:\s+(.*)\n/.exec(output);
      if (versionMatch) {
        return 'docker ' + versionMatch[1];
      }
      return undefined;
    });
}

function handleCommonErrors(error, targetImage: string) {
  if (error.indexOf('command not found') !== -1) {
    throw new Error('Snyk docker CLI was not found');
  }
  if (error.indexOf('Cannot connect to the Docker daemon') !== -1) {
    throw new Error('Cannot connect to the Docker daemon. Is the docker'
    + ' daemon running?');
  }
  const ERROR_LOADING_IMAGE_STR = 'Error loading image from docker engine:';
  if (error.indexOf(ERROR_LOADING_IMAGE_STR) !== -1) {
    if (error.indexOf('reference does not exist') !== -1) {
      throw new Error(
        `Docker image was not found locally: ${targetImage}`);
    }
    if (error.indexOf('permission denied while trying to connect') !== -1) {
      let errString = error.split(ERROR_LOADING_IMAGE_STR)[1];
      errString = (errString || '').slice(0, -2); // remove trailing \"
      throw new Error(
        'Permission denied connecting to docker daemon. ' +
        'Please make sure user has the required permissions. ' +
        'Error string: ' + errString);
    }
  }
  if (error.indexOf('Error getting docker client:') !== -1) {
    throw new Error('Failed getting docker client');
  }
  if (error.indexOf('Error processing image:') !== -1) {
    throw new Error('Failed processing image:' + targetImage);
  }
}

function getDependencies(targetImage: string) {
  let result;
  return analyzer.analyze(targetImage)
    .then((output) => {
      result = parseAnalysisResults(output);
      return buildTree(
        targetImage, result.type, result.depInfosList, result.targetOS);
    })
    .then((pkg) => {
      return {
        package: pkg,
        packageManager: result.type,
        imageId: result.imageId,
      };
    })
    .catch((error) => {
      if (typeof error === 'string') {
        debug(`Error while running analyzer: '${error}'`);
        handleCommonErrors(error, targetImage);
        let errorMsg = error;
        const errorMatch = /msg="(.*)"/g.exec(errorMsg);
        if (errorMatch) {
          errorMsg = errorMatch[1];
        }
        throw new Error(errorMsg);
      }
      throw error;
    });
}

function parseAnalysisResults(analysisJson) {
  const analysisResult = analysisJson.results.filter((res) => {
    return res.Analysis && res.Analysis.length > 0;
  })[0];

  if (!analysisResult) {
    throw new Error(
      'Failed to detect a supported Linux package manager (deb/rpm/apk)');
  }

  let depType;
  switch (analysisResult.AnalyzeType) {
    case 'Apt': {
      depType = 'deb';
      break;
    }
    default: {
      depType = analysisResult.AnalyzeType.toLowerCase();
    }
  }

  return {
    imageId: analysisJson.imageId,
    targetOS: analysisJson.osRelease,
    type: depType,
    depInfosList: analysisResult.Analysis,
  };
}

function buildTree(targetImage, depType, depInfosList, targetOS) {
  const targetSplit = targetImage.split(':');
  const imageName = targetSplit[0];
  const imageVersion = targetSplit[1] ? targetSplit[1] : 'latest';

  const root = {
    name: imageName,
    version: imageVersion,
    targetOS,
    packageFormatVersion: depType + ':0.0.1',
    dependencies: {},
  };

  const depsMap = depInfosList.reduce((acc, depInfo) => {
    const name = depInfo.Name;
    acc[name] = depInfo;
    return acc;
  }, {});

  const virtualDepsMap = depInfosList.reduce((acc, depInfo) => {
    const providesNames = depInfo.Provides || [];
    providesNames.forEach((name) => {
      acc[name] = depInfo;
    });
    return acc;
  }, {});

  const depsCounts = {};
  depInfosList.forEach((depInfo) => {
    countDepsRecursive(
      depInfo.Name, new Set(), depsMap, virtualDepsMap, depsCounts);
  });
  const DEP_FREQ_THRESHOLD = 100;
  const tooFrequentDepNames = Object.keys(depsCounts)
    .filter((depName) => {
      return depsCounts[depName] > DEP_FREQ_THRESHOLD;
    });

  const attachDeps = (depInfos) => {
    const depNamesToSkip = new Set(tooFrequentDepNames);
    depInfos.forEach((depInfo) => {
      const subtree = buildTreeRecurisve(
        depInfo.Name, new Set(), depsMap, virtualDepsMap, depNamesToSkip);
      if (subtree) {
        root.dependencies[subtree.name] = subtree;
      }
    });
  };

  // attach (as direct deps) pkgs not marked auto-installed:
  const manuallyInstalledDeps = depInfosList.filter((depInfo) => {
    return !depInfo.AutoInstalled;
  });
  attachDeps(manuallyInstalledDeps);

  // attach (as direct deps) pkgs marked as auto-insatalled,
  //  but not dependant upon:
  const notVisitedDeps = depInfosList.filter((depInfo) => {
    const depName = depInfo.Name;
    return !(depsMap[depName]._visited);
  });
  attachDeps(notVisitedDeps);

  // group all the "too frequest" deps under a meta package:
  if (tooFrequentDepNames.length > 0) {
    const tooFrequentDeps = tooFrequentDepNames.map((name) => {
      return depsMap[name];
    });

    const metaSubtree = {
      name: 'meta-common-packages',
      version: 'meta',
      dependencies: {},
    };

    tooFrequentDeps.forEach((depInfo) => {
      const pkg = {
        name: depFullName(depInfo),
        version: depInfo.Version,
      };
      metaSubtree.dependencies[pkg.name] = pkg;
    });

    root.dependencies[metaSubtree.name] = metaSubtree;
  }

  return root;
}

function buildTreeRecurisve(
  depName, ancestors, depsMap, virtualDepsMap, depNamesToSkip) {
  const depInfo = depsMap[depName] || virtualDepsMap[depName];
  if (!depInfo) {
    return null;
  }

  // "realName" as the argument depName might be a virtual pkg
  const realName = depInfo.Name;
  const fullName = depFullName(depInfo);
  if (ancestors.has(fullName) || depNamesToSkip.has(realName)) {
    return null;
  }

  const tree: {
    name: string;
    version: string;
    dependencies?: any;
  } = {
    name: fullName,
    version: depInfo.Version,
  };

  if (depInfo._visited) {
    return tree;
  }
  depInfo._visited = true;

  const newAncestors = (new Set(ancestors)).add(fullName);

  const deps = depInfo.Deps || {};
  Object.keys(deps).forEach((name) => {
    const subTree = buildTreeRecurisve(
      name, newAncestors, depsMap, virtualDepsMap, depNamesToSkip);
    if (subTree) {
      if (!tree.dependencies) {
        tree.dependencies = {};
      }
      tree.dependencies[subTree.name] = subTree;
    }
  });

  return tree;
}

function countDepsRecursive(
  depName, ancestors, depsMap, virtualDepsMap, depCounts) {
  const depInfo = depsMap[depName] || virtualDepsMap[depName];
  if (!depInfo) {
    return;
  }

  // "realName" as the argument depName might be a virtual pkg
  const realName = depInfo.Name;
  if (ancestors.has(realName)) {
    return;
  }

  depCounts[realName] = (depCounts[realName] || 0) + 1;

  const newAncestors = (new Set(ancestors)).add(realName);
  const deps = depInfo.Deps || {};
  Object.keys(deps).forEach((name) => {
    countDepsRecursive(
      name, newAncestors, depsMap, virtualDepsMap, depCounts);
  });
}

function depFullName(depInfo) {
  let fullName = depInfo.Name;
  if (depInfo.Source) {
    fullName = depInfo.Source + '/' + fullName;
  }
  return fullName;
}
