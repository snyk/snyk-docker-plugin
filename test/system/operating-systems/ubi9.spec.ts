import { scan } from "../../../lib/index";
import { execute } from "../../../lib/sub-process";

describe("redhat ubi9 tests", () => {
  afterAll(async () => {
    await execute("docker", ["image", "rm", "redhat/ubi9-micro:9.3"]).catch(
      () => {
        console.error(`tests teardown failed to remove docker image`);
      },
    );
  });

  it("should correctly analyze an ubi9 image by tag", async () => {
    const image = "redhat/ubi9-micro:9.3";
    const pluginResult = await scan({
      path: image,
      platform: "linux/amd64",
    });

    expect(pluginResult).toMatchSnapshot();
  });
});
