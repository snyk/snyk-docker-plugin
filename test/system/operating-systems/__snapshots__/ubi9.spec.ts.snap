// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`redhat ubi9 tests should correctly analyze an ubi9 image by tag 1`] = `
Object {
  "scanResults": Array [
    Object {
      "facts": Array [
        Object {
          "data": Object {
            "graph": Object {
              "nodes": Array [
                Object {
                  "deps": Array [
                    Object {
                      "nodeId": "basesystem@11-13.el9",
                    },
                    Object {
                      "nodeId": "bash@5.1.8-6.el9_1",
                    },
                    Object {
                      "nodeId": "coreutils-single@8.32-34.el9",
                    },
                    Object {
                      "nodeId": "filesystem@3.16-2.el9",
                    },
                    Object {
                      "nodeId": "glibc@2.34-83.el9_3.12",
                    },
                    Object {
                      "nodeId": "glibc-common@2.34-83.el9_3.12",
                    },
                    Object {
                      "nodeId": "glibc-minimal-langpack@2.34-83.el9_3.12",
                    },
                    Object {
                      "nodeId": "libacl@2.3.1-3.el9",
                    },
                    Object {
                      "nodeId": "libattr@2.5.1-3.el9",
                    },
                    Object {
                      "nodeId": "libcap@2.48-9.el9_2",
                    },
                    Object {
                      "nodeId": "libgcc@11.4.1-2.1.el9",
                    },
                    Object {
                      "nodeId": "libselinux@3.5-1.el9",
                    },
                    Object {
                      "nodeId": "libsepol@3.5-1.el9",
                    },
                    Object {
                      "nodeId": "ncurses-base@6.2-10.20210508.el9",
                    },
                    Object {
                      "nodeId": "ncurses-libs@6.2-10.20210508.el9",
                    },
                    Object {
                      "nodeId": "pcre2@10.40-2.el9",
                    },
                    Object {
                      "nodeId": "pcre2-syntax@10.40-2.el9",
                    },
                    Object {
                      "nodeId": "redhat-release@9.3-0.5.el9",
                    },
                    Object {
                      "nodeId": "setup@2.13.7-9.el9",
                    },
                    Object {
                      "nodeId": "tzdata@2024a-1.el9",
                    },
                  ],
                  "nodeId": "root-node",
                  "pkgId": "docker-image|redhat/ubi9-micro@9.3",
                },
                Object {
                  "deps": Array [],
                  "nodeId": "basesystem@11-13.el9",
                  "pkgId": "basesystem@11-13.el9",
                },
                Object {
                  "deps": Array [],
                  "nodeId": "bash@5.1.8-6.el9_1",
                  "pkgId": "bash@5.1.8-6.el9_1",
                },
                Object {
                  "deps": Array [],
                  "nodeId": "coreutils-single@8.32-34.el9",
                  "pkgId": "coreutils-single@8.32-34.el9",
                },
                Object {
                  "deps": Array [],
                  "nodeId": "filesystem@3.16-2.el9",
                  "pkgId": "filesystem@3.16-2.el9",
                },
                Object {
                  "deps": Array [],
                  "nodeId": "glibc@2.34-83.el9_3.12",
                  "pkgId": "glibc@2.34-83.el9_3.12",
                },
                Object {
                  "deps": Array [],
                  "nodeId": "glibc-common@2.34-83.el9_3.12",
                  "pkgId": "glibc-common@2.34-83.el9_3.12",
                },
                Object {
                  "deps": Array [],
                  "nodeId": "glibc-minimal-langpack@2.34-83.el9_3.12",
                  "pkgId": "glibc-minimal-langpack@2.34-83.el9_3.12",
                },
                Object {
                  "deps": Array [],
                  "nodeId": "libacl@2.3.1-3.el9",
                  "pkgId": "libacl@2.3.1-3.el9",
                },
                Object {
                  "deps": Array [],
                  "nodeId": "libattr@2.5.1-3.el9",
                  "pkgId": "libattr@2.5.1-3.el9",
                },
                Object {
                  "deps": Array [],
                  "nodeId": "libcap@2.48-9.el9_2",
                  "pkgId": "libcap@2.48-9.el9_2",
                },
                Object {
                  "deps": Array [],
                  "nodeId": "libgcc@11.4.1-2.1.el9",
                  "pkgId": "libgcc@11.4.1-2.1.el9",
                },
                Object {
                  "deps": Array [],
                  "nodeId": "libselinux@3.5-1.el9",
                  "pkgId": "libselinux@3.5-1.el9",
                },
                Object {
                  "deps": Array [],
                  "nodeId": "libsepol@3.5-1.el9",
                  "pkgId": "libsepol@3.5-1.el9",
                },
                Object {
                  "deps": Array [],
                  "nodeId": "ncurses-base@6.2-10.20210508.el9",
                  "pkgId": "ncurses-base@6.2-10.20210508.el9",
                },
                Object {
                  "deps": Array [],
                  "nodeId": "ncurses-libs@6.2-10.20210508.el9",
                  "pkgId": "ncurses-libs@6.2-10.20210508.el9",
                },
                Object {
                  "deps": Array [],
                  "nodeId": "pcre2@10.40-2.el9",
                  "pkgId": "pcre2@10.40-2.el9",
                },
                Object {
                  "deps": Array [],
                  "nodeId": "pcre2-syntax@10.40-2.el9",
                  "pkgId": "pcre2-syntax@10.40-2.el9",
                },
                Object {
                  "deps": Array [],
                  "nodeId": "redhat-release@9.3-0.5.el9",
                  "pkgId": "redhat-release@9.3-0.5.el9",
                },
                Object {
                  "deps": Array [],
                  "nodeId": "setup@2.13.7-9.el9",
                  "pkgId": "setup@2.13.7-9.el9",
                },
                Object {
                  "deps": Array [],
                  "nodeId": "tzdata@2024a-1.el9",
                  "pkgId": "tzdata@2024a-1.el9",
                },
              ],
              "rootNodeId": "root-node",
            },
            "pkgManager": Object {
              "name": "rpm",
              "repositories": Array [
                Object {
                  "alias": "rhel:9.3",
                },
              ],
            },
            "pkgs": Array [
              Object {
                "id": "docker-image|redhat/ubi9-micro@9.3",
                "info": Object {
                  "name": "docker-image|redhat/ubi9-micro",
                  "version": "9.3",
                },
              },
              Object {
                "id": "basesystem@11-13.el9",
                "info": Object {
                  "name": "basesystem",
                  "purl": "pkg:rpm/rhel/basesystem@11-13.el9?distro=rhel-9.3&repositories=rhel-9-for-x86_64-baseos-rpms%2Crhel-9-for-x86_64-appstream-rpms",
                  "version": "11-13.el9",
                },
              },
              Object {
                "id": "bash@5.1.8-6.el9_1",
                "info": Object {
                  "name": "bash",
                  "purl": "pkg:rpm/rhel/bash@5.1.8-6.el9_1?distro=rhel-9.3&repositories=rhel-9-for-x86_64-baseos-rpms%2Crhel-9-for-x86_64-appstream-rpms",
                  "version": "5.1.8-6.el9_1",
                },
              },
              Object {
                "id": "coreutils-single@8.32-34.el9",
                "info": Object {
                  "name": "coreutils-single",
                  "purl": "pkg:rpm/rhel/coreutils-single@8.32-34.el9?distro=rhel-9.3&repositories=rhel-9-for-x86_64-baseos-rpms%2Crhel-9-for-x86_64-appstream-rpms",
                  "version": "8.32-34.el9",
                },
              },
              Object {
                "id": "filesystem@3.16-2.el9",
                "info": Object {
                  "name": "filesystem",
                  "purl": "pkg:rpm/rhel/filesystem@3.16-2.el9?distro=rhel-9.3&repositories=rhel-9-for-x86_64-baseos-rpms%2Crhel-9-for-x86_64-appstream-rpms",
                  "version": "3.16-2.el9",
                },
              },
              Object {
                "id": "glibc@2.34-83.el9_3.12",
                "info": Object {
                  "name": "glibc",
                  "purl": "pkg:rpm/rhel/glibc@2.34-83.el9_3.12?distro=rhel-9.3&repositories=rhel-9-for-x86_64-baseos-rpms%2Crhel-9-for-x86_64-appstream-rpms",
                  "version": "2.34-83.el9_3.12",
                },
              },
              Object {
                "id": "glibc-common@2.34-83.el9_3.12",
                "info": Object {
                  "name": "glibc-common",
                  "purl": "pkg:rpm/rhel/glibc-common@2.34-83.el9_3.12?distro=rhel-9.3&repositories=rhel-9-for-x86_64-baseos-rpms%2Crhel-9-for-x86_64-appstream-rpms",
                  "version": "2.34-83.el9_3.12",
                },
              },
              Object {
                "id": "glibc-minimal-langpack@2.34-83.el9_3.12",
                "info": Object {
                  "name": "glibc-minimal-langpack",
                  "purl": "pkg:rpm/rhel/glibc-minimal-langpack@2.34-83.el9_3.12?distro=rhel-9.3&repositories=rhel-9-for-x86_64-baseos-rpms%2Crhel-9-for-x86_64-appstream-rpms",
                  "version": "2.34-83.el9_3.12",
                },
              },
              Object {
                "id": "libacl@2.3.1-3.el9",
                "info": Object {
                  "name": "libacl",
                  "purl": "pkg:rpm/rhel/libacl@2.3.1-3.el9?distro=rhel-9.3&repositories=rhel-9-for-x86_64-baseos-rpms%2Crhel-9-for-x86_64-appstream-rpms",
                  "version": "2.3.1-3.el9",
                },
              },
              Object {
                "id": "libattr@2.5.1-3.el9",
                "info": Object {
                  "name": "libattr",
                  "purl": "pkg:rpm/rhel/libattr@2.5.1-3.el9?distro=rhel-9.3&repositories=rhel-9-for-x86_64-baseos-rpms%2Crhel-9-for-x86_64-appstream-rpms",
                  "version": "2.5.1-3.el9",
                },
              },
              Object {
                "id": "libcap@2.48-9.el9_2",
                "info": Object {
                  "name": "libcap",
                  "purl": "pkg:rpm/rhel/libcap@2.48-9.el9_2?distro=rhel-9.3&repositories=rhel-9-for-x86_64-baseos-rpms%2Crhel-9-for-x86_64-appstream-rpms",
                  "version": "2.48-9.el9_2",
                },
              },
              Object {
                "id": "libgcc@11.4.1-2.1.el9",
                "info": Object {
                  "name": "libgcc",
                  "purl": "pkg:rpm/rhel/libgcc@11.4.1-2.1.el9?distro=rhel-9.3&repositories=rhel-9-for-x86_64-baseos-rpms%2Crhel-9-for-x86_64-appstream-rpms",
                  "version": "11.4.1-2.1.el9",
                },
              },
              Object {
                "id": "libselinux@3.5-1.el9",
                "info": Object {
                  "name": "libselinux",
                  "purl": "pkg:rpm/rhel/libselinux@3.5-1.el9?distro=rhel-9.3&repositories=rhel-9-for-x86_64-baseos-rpms%2Crhel-9-for-x86_64-appstream-rpms",
                  "version": "3.5-1.el9",
                },
              },
              Object {
                "id": "libsepol@3.5-1.el9",
                "info": Object {
                  "name": "libsepol",
                  "purl": "pkg:rpm/rhel/libsepol@3.5-1.el9?distro=rhel-9.3&repositories=rhel-9-for-x86_64-baseos-rpms%2Crhel-9-for-x86_64-appstream-rpms",
                  "version": "3.5-1.el9",
                },
              },
              Object {
                "id": "ncurses-base@6.2-10.20210508.el9",
                "info": Object {
                  "name": "ncurses-base",
                  "purl": "pkg:rpm/rhel/ncurses-base@6.2-10.20210508.el9?distro=rhel-9.3&repositories=rhel-9-for-x86_64-baseos-rpms%2Crhel-9-for-x86_64-appstream-rpms",
                  "version": "6.2-10.20210508.el9",
                },
              },
              Object {
                "id": "ncurses-libs@6.2-10.20210508.el9",
                "info": Object {
                  "name": "ncurses-libs",
                  "purl": "pkg:rpm/rhel/ncurses-libs@6.2-10.20210508.el9?distro=rhel-9.3&repositories=rhel-9-for-x86_64-baseos-rpms%2Crhel-9-for-x86_64-appstream-rpms",
                  "version": "6.2-10.20210508.el9",
                },
              },
              Object {
                "id": "pcre2@10.40-2.el9",
                "info": Object {
                  "name": "pcre2",
                  "purl": "pkg:rpm/rhel/pcre2@10.40-2.el9?distro=rhel-9.3&repositories=rhel-9-for-x86_64-baseos-rpms%2Crhel-9-for-x86_64-appstream-rpms",
                  "version": "10.40-2.el9",
                },
              },
              Object {
                "id": "pcre2-syntax@10.40-2.el9",
                "info": Object {
                  "name": "pcre2-syntax",
                  "purl": "pkg:rpm/rhel/pcre2-syntax@10.40-2.el9?distro=rhel-9.3&repositories=rhel-9-for-x86_64-baseos-rpms%2Crhel-9-for-x86_64-appstream-rpms",
                  "version": "10.40-2.el9",
                },
              },
              Object {
                "id": "redhat-release@9.3-0.5.el9",
                "info": Object {
                  "name": "redhat-release",
                  "purl": "pkg:rpm/rhel/redhat-release@9.3-0.5.el9?distro=rhel-9.3&repositories=rhel-9-for-x86_64-baseos-rpms%2Crhel-9-for-x86_64-appstream-rpms",
                  "version": "9.3-0.5.el9",
                },
              },
              Object {
                "id": "setup@2.13.7-9.el9",
                "info": Object {
                  "name": "setup",
                  "purl": "pkg:rpm/rhel/setup@2.13.7-9.el9?distro=rhel-9.3&repositories=rhel-9-for-x86_64-baseos-rpms%2Crhel-9-for-x86_64-appstream-rpms",
                  "version": "2.13.7-9.el9",
                },
              },
              Object {
                "id": "tzdata@2024a-1.el9",
                "info": Object {
                  "name": "tzdata",
                  "purl": "pkg:rpm/rhel/tzdata@2024a-1.el9?distro=rhel-9.3&repositories=rhel-9-for-x86_64-baseos-rpms%2Crhel-9-for-x86_64-appstream-rpms",
                  "version": "2024a-1.el9",
                },
              },
            ],
            "schemaVersion": "1.3.0",
          },
          "type": "depGraph",
        },
        Object {
          "data": "sha256:0a76a9bf80abadb332376f659186555601fe7d9d26979cf01d6441a8133e1af1",
          "type": "imageId",
        },
        Object {
          "data": Array [
            "sha256:f6e375c0d5ad6710500dc4fe0f0e6c5c43c65465c50ac3ce5c152cc9627f5134",
          ],
          "type": "imageLayers",
        },
        Object {
          "data": Object {
            "architecture": "x86_64",
            "build-date": "2024-02-29T16:04:33",
            "com.redhat.component": "ubi9-micro-container",
            "com.redhat.license_terms": "https://www.redhat.com/en/about/red-hat-end-user-license-agreements#UBI",
            "description": "Very small image which doesn't install the package manager.",
            "distribution-scope": "public",
            "io.buildah.version": "1.29.0",
            "io.k8s.description": "Very small image which doesn't install the package manager.",
            "io.k8s.display-name": "Ubi9-micro",
            "io.openshift.expose-services": "",
            "maintainer": "Red Hat, Inc.",
            "name": "ubi9/ubi-micro",
            "release": "15",
            "summary": "ubi9 micro image",
            "url": "https://access.redhat.com/containers/#/registry.access.redhat.com/ubi9/ubi-micro/images/9.3-15",
            "vcs-ref": "303dc144996db01765d69e8ea45d2d617d953e42",
            "vcs-type": "git",
            "vendor": "Red Hat, Inc.",
            "version": "9.3",
          },
          "type": "imageLabels",
        },
        Object {
          "data": "2024-02-29T16:07:21.552785209Z",
          "type": "imageCreationTime",
        },
        Object {
          "data": Array [
            "sha256:f6e375c0d5ad6710500dc4fe0f0e6c5c43c65465c50ac3ce5c152cc9627f5134",
          ],
          "type": "rootFs",
        },
        Object {
          "data": "Red Hat Enterprise Linux 9.3 (Plow)",
          "type": "imageOsReleasePrettyName",
        },
        Object {
          "data": Object {
            "names": Array [
              "redhat/ubi9-micro:9.3",
            ],
          },
          "type": "imageNames",
        },
      ],
      "identity": Object {
        "args": Object {
          "platform": "linux/amd64",
        },
        "type": "rpm",
      },
      "target": Object {
        "image": "docker-image|redhat/ubi9-micro",
      },
    },
  ],
}
`;
